cdm/core/src/main/java/ucar/nc2/dataset/VariableDS.java

```diff
@@ -261,23 +258,48 @@ Array convert(Array data, Set<NetcdfDataset.Enhance> enhancements) {
       if (this.isVariableLength) {
         return data;
       }
+      if (!dataType.isNumeric()) {
+        return data;
+      }
+      // datatype of the result depends on what enhancements were applied
+      DataType convertedType = data.getDataType();
+
+      // TODO: change to a provider for extensible Enhancements
+      List<Enhancement> toApply = new ArrayList<>();
       if (enhancements.contains(Enhance.ConvertUnsigned) && unsignedConversion != null) {
-        data = unsignedConversion.convertUnsigned(data);
+        toApply.add(unsignedConversion);
+        convertedType = unsignedConversion.getOutType();
       }
       if (enhancements.contains(Enhance.ApplyScaleOffset) && scaleOffset != null) {
-        data = scaleOffset.removeScaleOffset(data);
+        toApply.add(scaleOffset);
+        convertedType = scaleOffset.getScaledOffsetType();
       }
       if (enhancements.contains(Enhance.ConvertMissing) && convertMissing != null
           && (dataType == DataType.FLOAT || dataType == DataType.DOUBLE)) {
-        data = convertMissing.convertMissing(data);
+        toApply.add(convertMissing);
       }
       if (enhancements.contains(Enhance.ApplyStandardizer) && standardizer != null) {
-        data = standardizer.convert(data);
+        toApply.add(standardizer);
       }
       if (enhancements.contains(Enhance.ApplyNormalizer) && normalizer != null) {
-        data = normalizer.convert(data);
+        toApply.add(normalizer);
+      }
+
+      double[] dataArray = (double[]) data.get1DJavaArray(DataType.DOUBLE);
+
+      List<Double> list = Arrays.stream(dataArray).parallel().map((num) -> {
+        for (Enhancement e : toApply) {
+          num = e.convert(num);
+        }
+        return num;
+      }).boxed().collect(Collectors.toList());
+
+      Array out = Array.factory(convertedType, data.getShape());
+      IndexIterator iterOut = out.getIndexIterator();
+      for (int i = 0; i < list.size(); i++) {
+        iterOut.setObjectNext(list.get(i));
       }
-      return data;
+      return out;
```

Comment:
on second thought that only works if the `out` type is always double 🤔

Reply:
so maybe your solution is good as is except we don't want all three arrays in memory at the same time. So perhaps we can just do `dataArray = null` after creating `list` (or another solution, maybe make it go out of scope somehow)

Reply:
It may be better to still create the `double[]` from the stream with `toArray()` list I suggested above since then no boxing needs to happen. And then still iterate over `out` to create the final array of the correct type. Sorry for putting too many comments about this 😆

Reply:
`toArray` only turns the stream back into an array, it doesn't reference the mapped data. You have to `collect` to get the mapped values